### ISSUE TITLE
Different numfig_format settings for caption prefix and reference

### ISSUE THREAD

**Issue by asfg84** [sphinx-doc/sphinx#4086]
Since the numfig_format is used as caption prefix and also for referencing figures/tables etc., it is not possible to use different formats. I would like to have a reference in text as _"see Figure 1"_, and the caption of my figure should be 

_"Figure 1: Example image"_

Instead, with `numfig_format={'figure': 'Figure %s'}` I get 

_"Figure 1 Example image"_ (without colon)

The output in Latex pdf is correct, this only applies to the html output.

When I add the colon to the `numfig_format numfig_format={'figure': 'Figure %s:'}`, I also get the colon in the text reference, which I don't want to have. (And I have two colons in the caption of the latex output and no space between Figure and the number anymore).

### Environment info
- Sphinx version: 1.6.3

**Comment by misev** [comment]
+1, I guess this should be trivial to correct?

Also I came across a bug: if you change only the 'figure' format with `numfig_format={'figure': 'Figure %s'}`, and also have Tables for example in the document, you get this:
```
Exception occurred:
  File "/usr/lib/python2.7/dist-packages/sphinx/writers/html.py", line 279, in append_fignumber
    self.body.append(prefix % numsstr + ' ')
TypeError: not all arguments converted during string formatting
The full traceback has been saved in /tmp/sphinx-err-PjF2Qy.log, if you want to report the issue to the developers.
```
Specifying all formats fixed the problem:
```
numfig_format = {'figure': 'Figure %s', 'table': 'Table %s',
                 'code-block': 'Listing %s', 'section': 'Section %s'}
```

**Comment by tk0miya** [comment]
>+1, I guess this should be trivial to correct?

Really? We always welcome your contribution. Please send a PR for us :-)

On the other hand, second one is another bug. Could you file a new issue please?
I'll take a look later.

**Comment by misev** [comment]
> Really? We always welcome your contribution. Please send a PR for us :-)

I don't think it is very efficient for me to go into fixing something small (at least it looks so, can't be sure of course), given that so far I have zero experience contributing to this project :-)

New issue for the other problem: #4828

**Comment by tk0miya** [comment]
Thanks for filing. I'll take a look later.

About another case, I feel it is difficult to fix. So I can't promise you exactly when. Please challenge the problem If you are interested in :-)

**Comment by tesch1** [comment]
I'm interested in it :o)

Before I dive in, here's my proposal, what do you all think?

Adding a config setting `numfig_format_caption` that has the same options (a dict with keys 'figure', 'table', 'code-block' and 'section') and behavior as `numfig_format`, but its entries override the behavior of `numfig_format` only when generating output being used for a caption.

Any unset `numfig_format_caption` entries will default to the value of `numfig_format`'s

---

My first though was to create a setting that would just specify a separator, but this is probably about the same amount of work, and is more flexible, say, if someone wants their text to say "Figure x" but the figure captions to say "Fig. x:", which isn't so uncommon.

Or any other proposal?

**Comment by tesch1** [comment]
Also, the default for `numfig_format_caption` will be to include the ':' in the caption between the , as is done for the latex output.

Or, an alternative to that whole proposal would be to just add the ':' to the html output, it would force everyone to adopt that format, but it's an improvement for them anyway, right?  ;o)

**Comment by misev** [comment]
> Or, an alternative to that whole proposal would be to just add the ':' to the html output, it would force everyone to adopt that format, but it's an improvement for them anyway, right? ;o)

Technically this would be the fix for this bug, to simply make the html output consistent with the latex one.

Supporting customizable captions as you proposed in the previous comment would be considered a separate feature?

**Comment by tesch1** [comment]
I just found a workaround (/better way), add this to your custom css:
```
span.caption-number:after {
   content: " : ";
}
```

**Comment by mishall-ko** [comment]
I made captions with numfig_format in document. 
I want to make texts written with numfig_format as bold in document. How can I add options in the following scripts of conf.py?
Figure 1.1 Example of xx     ----->

**Figure 1.1 Example of xx**
==========================================
numfig_format = {'figure': 'Figure %s', 'table': 'Table %s',
                 'code-block': 'Listing %s', 'section': 'Section %s'}

**Comment by wrobell** [comment]
For my documentation I am setting `numfig_format` figure setting to `Figure %s`. I need to use the following custom CSS
```
span.caption-number:after {
   content: ": ";
   margin-left: -0.25em;
}
```
The reason for the margin correction is because Sphinx generates an additional space
```
<span class="caption-number">Figure 1 </span>
```
IMHO, `Figure 1: Some caption` looks better than `Figure 1 : Some caption`.

It would be great if the additional space is removed from the HTML output.

To summarize
1. Use `numfig_format` to switch between `Fig. 1` and `Figure 1`.
2. Control punctuation and spacing between figure label and its caption with CSS `after` selector.
3. For above to work seamlessly, the Sphinx would need to stop adding the additional space in the caption number `span` tag.

Would that allow to close this ticket?
